pkg/tenancy/tenancy.go

```diff
@@ -136,3 +138,97 @@ func GetTenantFromGRPCMetadata(ctx context.Context) (string, bool) {
 	}
 	return md.Get(DefaultTenantHeader)[0], true
 }
+
+func EnforceQueryTenancy(tenantLabel string, tenant string, query string) (string, error) {
+	labelMatcher := &labels.Matcher{
+		Name:  tenantLabel,
+		Type:  labels.MatchEqual,
+		Value: tenant,
+	}
+
+	e := injectproxy.NewEnforcer(false, labelMatcher)
+
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		return "", errors.Wrap(err, "error parsing query string, when enforcing tenenacy")
+	}
+
+	if err := e.EnforceNode(expr); err != nil {
+		return "", errors.Wrap(err, "error enforcing label")
+	}
+
+	return expr.String(), nil
+}
+
+func getLabelMatchers(formMatchers []string, tenant string, enforceTenancy bool, tenantLabel string) ([][]*labels.Matcher, error) {
+	tenantLabelMatcher := &labels.Matcher{
+		Name:  tenantLabel,
+		Type:  labels.MatchEqual,
+		Value: tenant,
+	}
+
+	matcherSets := make([][]*labels.Matcher, 0, len(formMatchers))
+
+	// If tenancy is enforced, but there are no matchers at all, add the tenant matcher
+	if len(formMatchers) == 0 && enforceTenancy {
+		var matcher []*labels.Matcher
+		matcher = append(matcher, tenantLabelMatcher)
+		matcherSets = append(matcherSets, matcher)
+		return matcherSets, nil
+	}
+
+	for _, s := range formMatchers {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return nil, err
+		}
+
+		if enforceTenancy {
+			e := injectproxy.NewEnforcer(false, tenantLabelMatcher)
+			matchers, err = e.EnforceMatchers(matchers)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		matcherSets = append(matcherSets, matchers)
+	}
+
+	return matcherSets, nil
+}
+
+// This function will:
+// - Get tenant from HTTP header and add it to context.
+// - if tenancy is enforce, add a tenant matcher.
+func RewritePromQL(ctx context.Context, r *http.Request, tenantHeader string, defaultTenantID string, certTenantField string, enforceTenancy bool, tenantLabel string, queryStr string) (string, string, context.Context, error) {
+	tenant, err := GetTenantFromHTTP(r, tenantHeader, defaultTenantID, certTenantField)
+	if err != nil {
+		return "", "", ctx, err
+	}
+	ctx = context.WithValue(ctx, TenantKey, tenant)
+
+	if enforceTenancy {
```

Comment:
Do we need to pass this bool here and check? I think it might be cleaner to just check the bool in the `query/queryRange` method directly & only call `RewritePromQL` if it is true there.